### DISCUSSION TITLE
Implementing repeat opcode in Sleigh

### DISCUSSION THREAD

**Discussion by lab313ru** [NationalSecurityAgency/ghidra#5584] [Q&A]
I have an instruction which looks like the following: `rep reg1`, where `reg1` is repeats count. This function executes next instruction `reg1` times. For example, if I have the following code flow:

```
move #3, d0
move #1, d1
rep d0
move d1, (a0)+
```

it should be interpreted like this:
```
int off = 0;
int d1 = 1;
for (int d0 = 0; d0 < 3; ++d0) {
  a0[off++] = 1;
}
```

How can I implement this in sleigh? I tried to use something like this, but it didn't help:
![image](https://github.com/NationalSecurityAgency/ghidra/assets/7189309/36b13108-ce97-4382-adc2-883ed9476647)

But got the following:
![image](https://github.com/NationalSecurityAgency/ghidra/assets/7189309/70fa20f1-c78c-4820-8462-98b4f4e40b28)

**Comment by GhidorahRex** [comment]
You might want to use context values here. See how the REP prefix works in x86. Also the TI_MSP430X has a repeat context you can look at as well, although it's more convoluted and includes pcode for every instruction that might repeat that could probably be short-circuited by checking if there was an extension word to begin with.

The `rep` instruction would set the repeat context flag and set a context register and copy the register value to a dummy counter. If the actual register is being decremented, it's slightly different - you would assign a register value to the context.
a `^instruction` instruction would check if the repeat flag is set, and if it is,check and decrement the counter.

In sleigh pseudo-code, they would look like this, which I think would work

Dummy register:
```
define register offset=0xf00d size=4 [ COUNT ];
...
define context contextreg
repeat=(0,0) noflow
phase=(31,31)
;

:^instruction is phase=0 & repeat=1 & instruction [phase=1;] { COUNT = COUNT - 1; build instruction; if (COUNT  > 0) goto inst_start; }
:^instruction is phase=0 & repeat=0 & instruction [phase=1;] { build instruction; }

:rep reg is opcode=0xbeef & reg [repeat=1; globalset(inst_next, repeat); ]  { COUNT = reg; }
```

Context register:
```
define context contextreg
repeat=(0,0) noflow
repreg = (1,5) noflow
phase=(31,31)
;
...
define token optoken (32)
  reg = (0,4)
  regc = (0,4)
...
attach variables [ reg repreg ] [register list...];
...
:^instruction is phase=0 & repeat=1 & repreg & instruction [phase=1;] { repreg = repreg - 1; build instruction; if (repreg  > 0) goto inst_start; }
:^instruction is phase=0 & repeat=0 & instruction [phase=1;] { build instruction; }
...
:rep reg is opcode=0xbeef & reg & regc [repeat=1; globalset(inst_next, repeat);  repreg=regc; globalset(inst_next, repreg); ]  { }
```
